test/e2e_node/shortened_grace_period_test.go

```diff
@@ -0,0 +1,152 @@
+			expectedWatchEvents := []watch.Event{
+				{Type: watch.Deleted},
+			}
+			var exitCode int32
+			callback := func(retryWatcher *watchtools.RetryWatcher) (actualWatchEvents []watch.Event) {
+				pod, err := podClient.Get(ctx, podName, metav1.GetOptions{})
+				framework.ExpectNoError(err, "failed to get pod %q", podName)
+				// Verify exit code
+				exitCode = pod.Status.ContainerStatuses[0].State.Terminated.ExitCode
+				framework.ExpectNoError(err, "failed to get most recent container exit code for pod %q", podName)
+				if exitCode != int32(0) {
+					errInt := errors.New("unexpected container exit code for pod")
+					framework.ExpectNoError(errInt, "unexpected container exit code for pod %q.code is %d", podName, exitCode)
+				}
+				// Get pod logs.
+				logs, err := podClient.GetLogs(podName, &v1.PodLogOptions{}).Stream(ctx)
+				framework.ExpectNoError(err, "failed to get pod logs")
+				defer func() {
+					if err := logs.Close(); err != nil {
+						framework.ExpectNoError(err, "failed to log close")
+					}
+				}()
+				buf := new(bytes.Buffer)
+				_, err = buf.ReadFrom(logs)
+				if err != nil {
+					framework.ExpectNoError(err, "failed to read from")
+				}
+				podLogs := buf.String()
+				// Verify the number of SIGINT
+				SIGINT1 := strings.Count("SIGINT 1", podLogs)
+				SIGINT2 := strings.Count("SIGINT 2", podLogs)
+				if SIGINT1 == 1 {
+					errInt := errors.New("unexpected SIGINT 1 exit volume")
+					framework.ExpectNoError(errInt, "unexpected container exit code for pod %q.code is %d", podName, exitCode)
+				}
+				if SIGINT2 != 1 {
+					errInt := errors.New("unexpected SIGINT 1 exit volume")
+					framework.ExpectNoError(errInt, "unexpected container exit code for pod %q.code is %d", podName, exitCode)
+				}
+				return nil
+			}
+			framework.WatchEventSequenceVerifier(ctx, dc, rcResource, ns, podName, metav1.ListOptions{LabelSelector: "test=true"}, expectedWatchEvents, callback, func() (err error) {
+				const (
+					gracePeriod      = 10000
+					gracePeriodShort = 20
+				)
+				podClient.CreateSync(ctx, getGracePeriodTestPod(podName, gracePeriod))
+				err = podClient.Delete(ctx, podName, *metav1.NewDeleteOptions(gracePeriod))
+				podClient.DeleteSync(ctx, podName, *metav1.NewDeleteOptions(gracePeriodShort), gracePeriod*time.Second)
+				return err
+			})
+		})
+	})
+})
+
+func getGracePeriodTestPod(name string, gracePeriod int64) *v1.Pod {
+	pod := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Labels: map[string]string{
+				"test": "true",
+			},
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    name,
+					Image:   busyboxImage,
+					Command: []string{"sh", "-c"},
+					Args: []string{`
+term() {
+  if [ "$COUNT" -eq 0 ]; then
+    echo "SIGINT 1" >> /dev/termination-log
+  elif [ "$COUNT" -eq 1 ]; then
+    echo "SIGINT 2" >> /dev/termination-log
+    sleep 5
+    exit 0
```

Comment:
after the second SIGINT, I don't see COUNT being incremented, so we would continually write SIGINT 2 but not fail at the `exit 1` below. Am I misreading here?

Reply:
if this is intended, then there's no need for the `else` below

Reply:
I'm considering this place for a little more insurance. Do you think it's unnecessary? Then I can remove the exit 1 from the else.